consensus/hotstuff/timeoutcollector/timeout_processor.go

```diff
@@ -0,0 +1,292 @@
+		return fmt.Errorf("could not process invalid signature: %w", err)
+	}
+
+	if p.partialTCTracker.Track(totalWeight) {
+		p.notifier.OnPartialTcCreated(p.view)
+	}
+
+	// checking of conditions for building TC are satisfied
+	// At this point, we have enough signatures to build a TC. Another routine
+	// might just be at this point. To avoid duplicate work, only one routine can pass:
+	if !p.tcTracker.Track(totalWeight) {
+		return nil
+	}
+
+	tc, err := p.buildTC()
+	if err != nil {
+		return fmt.Errorf("internal error constructing TC: %w", err)
+	}
+	p.notifier.OnTcConstructedFromTimeouts(tc)
+
+	return nil
+}
+
+// validateTimeout performs validation of timeout object, verifies if timeout is correctly structured
+// and included QC and TC is correctly structured and signed.
+// ATTENTION: this function doesn't check if timeout signature is valid, this check happens in signature aggregator
+// Expected error returns during normal operations:
+// * model.InvalidTimeoutError - submitted invalid timeout(invalid structure or invalid signature)
+// * model.ErrViewForUnknownEpoch if no epoch containing the given view is known
+// All other errors should be treated as exceptions.
+func (p *TimeoutProcessor) validateTimeout(timeout *model.TimeoutObject) error {
+	// 1. check if it's correctly structured
+	// (a) Every TO must contain a QC
+	if timeout.NewestQC == nil {
+		return model.NewInvalidTimeoutErrorf(timeout, "TimeoutObject without QC is invalid")
+	}
+
+	if timeout.View < timeout.NewestQC.View {
+		return model.NewInvalidTimeoutErrorf(timeout, "TO's QC %d cannot be newer than the TO's view %d",
+			timeout.NewestQC.View, timeout.View)
+	}
+
+	// (b) If a TC is included, the TC must be for the past round, no matter whether a QC
+	//     for the last round is also included. In some edge cases, a node might observe
+	//     _both_ QC and TC for the previous round, in which case it can include both.
+	if timeout.LastViewTC != nil {
+		if timeout.View != timeout.LastViewTC.View+1 {
+			return model.NewInvalidTimeoutErrorf(timeout, "invalid TC for previous round")
+		}
+		if timeout.NewestQC.View < timeout.LastViewTC.NewestQC.View {
+			return model.NewInvalidTimeoutErrorf(timeout, "timeout.NewestQC has older view that the QC in timeout.LastViewTC")
+		}
+	}
+	// (c) The TO must contain a proof that sender legitimately entered timeout.View. Transitioning
+	//     to round timeout.View is possible either by observing a QC or a TC for the previous round.
+	//     If no QC is included, we require a TC to be present, which by check (1b) must be for
+	//     the previous round.
+	lastViewSuccessful := timeout.View == timeout.NewestQC.View+1
+	if !lastViewSuccessful {
+		// The TO's sender did _not_ observe a QC for round timeout.View-1. Hence, it should
+		// include a TC for the previous round. Otherwise, the TO is invalid.
+		if timeout.LastViewTC == nil {
+			return model.NewInvalidTimeoutErrorf(timeout, "timeout must include TC")
+		}
```

Comment:
I would suggest to move the [check on lines 193-195](https://github.com/onflow/flow-go/pull/2620/files#diff-88341aa1d744505d6d69c70fb72b2f2ec2ceda5bd61c69de860e145c4dc6dbfdR193-R195) here:
* it's easier to verify `(c)` when all the relevant checks are together
* it's the same structure as we use for with [similar validation with proposals](https://github.com/onflow/flow-go/blob/9f9d03bdd0ef107ed565dfab00e1aae7879ff1fb/consensus/hotstuff/validator/validator.go#L225-L228)

I'm also wondering, why do we not define this validation function in `Validator` alongside eg `ValidateQC`? To me it makes more sense for timeout validation to live alongside validation of the other entities.

Reply:
this block of code is only executed if the QC was _not_ for the previous round: https://github.com/onflow/flow-go/blob/f8ca29d13a8a4d2829de244c89810450f1462568/consensus/hotstuff/timeoutcollector/timeout_processor.go#L204-L205
In case the the last view _was_ successful (i.e. `lastViewSuccessful == true`), we would still want to check the TC. But we would just ignore the TC in this case, which I am concerned about from a BFT perspective. 

I honestly feel the current logic is correct (though I admit not the most intuitive):
(a) if a TC is included, it must be for the last view (independently of the QC)
(b) if the QC is not for the prior view, the TC cannot be nil 

The important aspect is that condition (a) must be universally enforced.

Reply:
Just to clarify, I agree the current logic is correct - my suggestion to move the checks around is about readability.

Reply:
🙈 Sorry, probably misunderstood. I only wanted to highlight that moving [lines 193 - 195](https://github.com/onflow/flow-go/blob/f8ca29d13a8a4d2829de244c89810450f1462568/consensus/hotstuff/timeoutcollector/timeout_processor.go#L192-L199) _inside_ the body of the [`if !lastViewSuccessful` statement](https://github.com/onflow/flow-go/blob/f8ca29d13a8a4d2829de244c89810450f1462568/consensus/hotstuff/timeoutcollector/timeout_processor.go#L205-L211) would create an additional edge case that is now properly handled.